sweep/sweeper.go

```diff
@@ -381,6 +382,30 @@ func (s *UtxoSweeper) Start() error {
 		defer s.wg.Done()
 
 		s.collector(blockEpochs.Epochs)
+
+		// The collector exited and won't longer handle incoming
+		// requests. This can happen on shutdown, when the block
+		// notifier shuts down before the sweeper and its clients. In
```

Comment:
In which case does the block notifier shut down? Aren't sweeper clients listen to the same quit channel?

Reply:
When lnd is signalled to shut down, the notifier will be stopped. This means the the `blockEpoch` channel will close and the `collector` goroutine in the sweeper will exit.

This all happens before the sweeper's `quit` channel has been closed, and any of the subsystems attemping to sweep an input at this point will block which stalls lnd shut down.

In the case I saw the `contractcourt` resolvers would attempt to sweep inputs, but since the notifier already had exited, the `collector` go routine wouldn't pick up their request. This caused shutdown to deadlock, since the `contractcourt` subsystems are stopped before the sweeper, and they would never exit.